learning_observer/learning_observer/lms_integration.py

```diff
@@ -0,0 +1,168 @@
+    '''
+    async def caller(request, **kwargs):
+        '''
+        Make an AJAX request to Google
+        '''
+        return await raw_ajax(request, remote_url, **kwargs)
+    setattr(caller, "__qualname__", name)
+
+    return caller
+
+def api_docs_handler(request, ENDPOINTS):
+    '''
+    Return a list of available endpoints.
+
+    Eventually, we should also document available function calls
+    '''
+    response = "URL Endpoints:\n\n"
+    for endpoint in ENDPOINTS:
+        response += f"{endpoint._local_url()}\n"
+        cleaners = endpoint._cleaners()
+        for c in cleaners:
+            response += f"   {cleaners[c]['local_url']}\n"
+    response += "\n\n Globals:"
+    return aiohttp.web.Response(text=response)
+
+def register_cleaner(data_source, cleaner_name, ENDPOINTS):
+    '''
+    This will register a cleaner function, for export both as a web service
+    and as a local function call.
+    '''
+    def decorator(f):
+        found = False
+        for endpoint in ENDPOINTS:
+            if endpoint.name == data_source:
+                found = True
+                endpoint._add_cleaner(
+                    cleaner_name,
+                    {
+                        'function': f,
+                        'local_url': f'{endpoint._local_url()}/{cleaner_name}',
+                        'name': cleaner_name
+                    }
+                )
+
+        if not found:
+            raise AttributeError(f"Data source {data_source} invalid; not found in endpoints.")
+        return f
+
+    return decorator
+
+def make_ajax_raw_handler(raw_canvas_ajax, remote_url):
+        async def ajax_passthrough(request):
+            runtime = learning_observer.runtime.Runtime(request)
+            response = await raw_canvas_ajax(runtime, remote_url, retry=True, **request.match_info)
+            return aiohttp.web.json_response(response)
+        return ajax_passthrough
+
+def make_cleaner_handler(raw_function, cleaner_function, name=None):
+    async def cleaner_handler(request):
+        response = cleaner_function(await raw_function(request, **request.match_info))
+        if isinstance(response, dict) or isinstance(response, list):
+            return aiohttp.web.json_response(response)
+        elif isinstance(response, str):
+            return aiohttp.web.Response(text=response)
+        else:
+            raise AttributeError(f"Invalid response type: {type(response)}")
+    if name is not None:
+        setattr(cleaner_handler, "__qualname__", name + "_handler")
+
+    return cleaner_handler
+
+def make_cleaner_function(raw_function, cleaner_function, name=None):
+    async def cleaner_local(request, **kwargs):
+        canvas_response = await raw_function(request, **kwargs)
+        clean = cleaner_function(canvas_response)
+        return clean
+    if name is not None:
+        setattr(cleaner_local, "__qualname__", name)
+    return cleaner_local
+
+class BaseLMS:
```

Comment:
Look at `kvs.py` for how the class naming and functionality ought to take place. Its best to use similar patterns to what is already in the system.
Rename to `_LMS`. See [this post](https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name) for information about underscore prefixed items in python.
Additionally, look at the `KVS_MAP`. We eventually want something similar so we can more quickly handle mapping a teacher to their appropriate LMS. This is not a priority, but worth noting.

Reply:
I changed this to LMS and not _LMS because it was referenced outside that file